plugins/synaptics-rmi/fu-synaptics-rmi-device.c

```diff
@@ -0,0 +1,1098 @@
+/*
+ * Copyright (C) 2012-2014 Andrew Duggan
+ * Copyright (C) 2012-2019 Synaptics Inc.
+ * Copyright (C) 2019 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fcntl.h>
+#include <errno.h>
+#include <linux/hidraw.h>
+#include <sys/ioctl.h>
+
+#include <glib/gstdio.h>
+
+#include "fu-io-channel.h"
+
+#include "fu-synaptics-rmi-common.h"
+#include "fu-synaptics-rmi-firmware.h"
+#include "fu-synaptics-rmi-v5-device.h"
+#include "fu-synaptics-rmi-v6-device.h"
+#include "fu-synaptics-rmi-v7-device.h"
+
+#define RMI_WRITE_REPORT_ID			0x9	/* output report */
+#define RMI_READ_ADDR_REPORT_ID			0xa	/* output report */
+#define RMI_READ_DATA_REPORT_ID			0xb	/* input report */
+#define RMI_ATTN_REPORT_ID			0xc	/* input report */
+#define RMI_SET_RMI_MODE_REPORT_ID		0xf	/* feature report */
+
+#define HID_RMI4_REPORT_ID			0
+#define HID_RMI4_READ_INPUT_COUNT		1
+#define HID_RMI4_READ_INPUT_DATA		2
+#define HID_RMI4_READ_OUTPUT_ADDR		2
+#define HID_RMI4_READ_OUTPUT_COUNT		4
+#define HID_RMI4_WRITE_OUTPUT_COUNT		1
+#define HID_RMI4_WRITE_OUTPUT_ADDR		2
+#define HID_RMI4_WRITE_OUTPUT_DATA		4
+#define HID_RMI4_FEATURE_MODE			1
+#define HID_RMI4_ATTN_INTERUPT_SOURCES		1
+#define HID_RMI4_ATTN_DATA			2
+
+typedef struct
+{
+	gint			 fd;
+	FuSynapticsRmiFlash	 flash;
+	GPtrArray		*functions;
+	FuIOChannel		*io_channel;
+	FuSynapticsRmiFunction	*f01;
+	FuSynapticsRmiFunction	*f34;
+	guint8			 page;
+	guint			 num_interrupt_regs;
+	guint8			 manufacturer_id;
+	guint8			 has_lts;
+	guint8			 has_sensor_id;
+	guint8			 has_query42;
+	guint8			 has_dds4_queries;
```

Comment:
For all these booleans, flags or bitmap maybe?

Reply:
I've converted them to all be `gboolean`s. I think considering we're going to have one touchpad, so we can afford wasting 15 bytes :)